File: rust/do/src/nightup/nightup.rs
use ini::Ini;
use std::env;
use std::process::ExitCode;

mod zig;

const HELP_MSG: &str = "
Usage:
    do.exe nightup zig
    do.exe nightup odin
    do.exe nightup v
    do.exe nightup go";

pub fn run(args: &[String]) -> ExitCode {
    if args.len() == 1 && (args[0] == "-h" || args[0] == "--help") {
        println!("{}", HELP_MSG);
        return ExitCode::SUCCESS;
    }
    if args.len() != 1 {
        eprintln!("{}", HELP_MSG);
        return ExitCode::FAILURE;
    }

    let base_home = match env::home_dir() {
        Some(path) => path,
        None => {
            eprintln!("impossible to get your home dir");
            return ExitCode::FAILURE;
        }
    };

    let mut download_path = base_home.clone();
    download_path.push("Downloads");
    let download_dir = match download_path.into_os_string().into_string() {
        Ok(s) => s,
        Err(_) => {
            eprintln!("download path is not valid UTF-8");
            return ExitCode::FAILURE;
        }
    };

    let mut ini_path = base_home;
    ini_path.push(".nightup");

    let conf = match Ini::load_from_file_noescape(&ini_path) {
        Ok(ini) => ini,
        Err(_) => {
            eprintln!("failed to open: {}", ini_path.display());
            return ExitCode::FAILURE;
        }
    };

    let section = match conf.section(Some("Windows")) {
        Some(section) => section,
        None => {
            eprintln!(r#"nightup ini: not found section: "Windows""#);
            return ExitCode::FAILURE;
        }
    };

    match args[0].as_str() {
        "zig" => {
            let dist_dir = match section.get("zig") {
                Some(path) => path,
                None => {
                    eprintln!(r#"nightup ini: not found path: "zig""#);
                    return ExitCode::FAILURE;
                }
            };
            zig::run(dist_dir, &download_dir)
        }
        _ => {
            eprintln!("nightup: unknown command '{}'\n{}", args[0], HELP_MSG);
            ExitCode::FAILURE
        }
    }
}

File: nim/do/src/nightup.nim
import std/[os, strformat, parsecfg]

import ./[utils]

import ./nightup/[
  zig,
  ]


const HELP_MSG = """
USAGE:
    do.exe nightup go
    do.exe nightup odin
    do.exe nightup v
    do.exe nightup zig
    do.exe nightup vim
OPTION:
    -h, --help                 ヘルプメッセージを表示"""

proc getDistDir(cfg: Config, key: string): string =
  result = getSectionValue(cfg, "Windows", key)
  if result == "":
    stderrMsgAndExit fmt"""nightup ini: not found path: "{key}""""

proc run*(args: seq[string]) =
  if args.len == 1 and (args[0] == "-h" or args[0] == "--help"):
    stdoutMsgAndExit HELP_MSG
  if args.len != 1:
    stderrMsgAndExit HELP_MSG

  # 1. ホームディレクトリの取得
  let homeDir = getHomeDir()
  if homeDir == "":
    stderrMsgAndExit "Impossible to get your home dir!"

  # パスの結合 ($HOME/.nightup)
  let iniPath = homeDir / ".nightup"

  # 2. INIファイルのロード
  var cfg: Config
  try:
    cfg = loadConfig(iniPath)
  except CatchableError:
    stderrMsgAndExit fmt"failed to open: {iniPath}"

  # 3. 一時保存ディレクトリの設定
  var downloadDir = getTempDir()
  if downloadDir == "":
    stderrMsgAndExit "failed to getTempDIr"

  # 4. 各言語のアップデート処理への振り分け
  case args[0]
  of "zig":
    let distDir = getDistDir(cfg, "zig")
    zig.run(distDir, downloadDir)
  # of "odin":
  #   let distDir = getDistDir("odin")
  #   # odinlang.run(distDir, downloadDir)
  # of "v":
  #   let distDir = getDistDir("v")
  #   # vlang.run(distDir, downloadDir)
  # of "go":
  #   let distDir = getDistDir("go")
  #   # golang.run(distDir, downloadDir)
  # of "vim":
  #   # vim.run()
  #   discard
  else:
    stderrMsgAndExit fmt"nightup: unknown command '{args[0]}'\n{HELP_MSG}"

when isMainModule:
  run(commandLineParams())
